### DISCUSSION TITLE
prisma2 on delete cascade...

### DISCUSSION THREAD

**Discussion by kdhyo** [prisma/prisma#3508] [Q&A] [answered]
https://www.prisma.io/docs/guides/database-workflows/cascading-deletes/mysql
https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-sql-node-mysql

I looked at the site above and referenced it.
but, "The change you are trying to make would violate the required relation 'PostToUser' between the `Post` and `User` models."
I want to delete All the referenced record and the referenced record..

**Comment by ryands17** [comment, marked as answer]
Could you share the SQL that you created and your `schema.prisma` after introspecting?

**Reply by kdhyo** [reply]
![prisma1](https://user-images.githubusercontent.com/60608509/92114239-7ad4eb00-ee2b-11ea-980d-5abcf31c8cc9.PNG)
![prisma2](https://user-images.githubusercontent.com/60608509/92114242-7c061800-ee2b-11ea-9b75-24633ef21f72.PNG)
![prisma3](https://user-images.githubusercontent.com/60608509/92114245-7c061800-ee2b-11ea-898e-260f51264692.PNG)
------
![prisma4](https://user-images.githubusercontent.com/60608509/92114246-7c9eae80-ee2b-11ea-94ea-346f16470468.PNG)
![prisma5](https://user-images.githubusercontent.com/60608509/92114643-241be100-ee2c-11ea-97c7-8136c7d1c1bd.PNG)


I tried two methods but it failed.
Please understand that you are using a translator because you are not good at English..

**Reply by ryands17** [reply]
@kdhyo 
1. Create the following tables in MySQL

```sql
create table `User` (
  `id` int auto_increment primary key,
  `name` varchar(256)
);

create table `Post` (
  `id` int auto_increment primary key,
  `title` varchar(256),
  `author_id` int,
  constraint `author_fk` foreign key (`author`) references `User`(`id`) on delete cascade
);

create table `Profile` (
  `id` int auto_increment primary key,
  `userId` int unique,
  constraint `profile_fk` foreign key (`userId`) references `User`(`id`) on delete cascade
);
```

2. Run the following command from the command line in your project

```
prisma introspect --experimental-reintrospection
```

3. Now your `schema.prisma` should look like this

```prisma
model Post {
  id        Int     @default(autoincrement()) @id
  title     String?
  author_id Int?
  User      User?   @relation(fields: [author_id], references: [id])

  @@index([author_id], name: "author_fk")
}

model Profile {
  id     Int   @default(autoincrement()) @id
  userId Int?  @unique
  User   User? @relation(fields: [userId], references: [id])
}

model User {
  id      Int      @default(autoincrement()) @id
  name    String?
  Post    Post[]
  Profile Profile?
}
```

Now if you perform a cascade delete, it will work fine.

For e.g.

```ts
const user = await prisma.user.create({
    data: {
      name: 'user 1',
      Post: {
        create: [{ title: 'post 1' }, { title: 'post 2' }],
      },
    },
  })

  await prisma.user.delete({
    where: { id: user.id },
  })
```

**Reply by kdhyo** [reply]
It was solved thanks to you!!!
.
If not excuse me I have a question.
- At the time of design, is it deleted due to 'on delete cascade'?
- What is the'?' after schema.prisma?

**Reply by ryands17** [reply]
Yes it is at the time of design.
The `?` means that the field is optional. And this was missing in your code and which was causing the error for cascade deletes. This `?` is required for the moment as there is an issue here https://github.com/prisma/prisma/issues/2057 that describes the same.

The `?` will not be required after the above issue is fixed.

You can read more about data modelling in Prisma [here](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/data-model/)

**Reply by kdhyo** [reply]
thank you so much!!
prisma team the best!
We will continue to use Prisma and contribute to our development!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
